### PR TITLE
[C++ verification] adjust lambda operator

### PR DESCRIPTION
**Pull request by XLiZHI**
Here is AST for lambda expr:
```c
| |       `-LambdaExpr 0x6f82f10 <col:28, line:22:9> '(lambda at main6.cpp:19:28)'
| |         |-CXXRecordDecl 0x6f82a78 <line:19:28> col:28 implicit class definition
| |         | |-DefinitionData lambda pass_in_registers standard_layout trivially_copyable literal can_const_default_init
| |         | | |-DefaultConstructor
| |         | | |-CopyConstructor simple trivial has_const_param needs_implicit implicit_has_const_param
| |         | | |-MoveConstructor exists simple trivial needs_implicit
| |         | | |-CopyAssignment trivial has_const_param needs_implicit implicit_has_const_param
| |         | | |-MoveAssignment
| |         | | `-Destructor simple irrelevant trivial
| |         | |-CXXMethodDecl 0x6f82bc0 <col:35, line:22:9> line:19:28 used constexpr operator() 'void () const' inline
| |         | | `-CompoundStmt 0x6f82d18 <line:20:9, line:22:9>
| |         | |   `-UnaryOperator 0x6f82d00 <line:21:13, col:17> 'int' postfix '++'
| |         | |     `-MemberExpr 0x6f82cd0 <col:13> 'int' lvalue ->data 0x6f828a8
| |         | |       `-CXXThisExpr 0x6f82cc0 <col:13> 'classB *' implicit this
| |         | |-FieldDecl 0x6f82e98 <line:19:29> col:29 implicit 'classB *'
| |         | `-CXXDestructorDecl 0x6f82f40 <col:28> col:28 implicit referenced ~(lambda at main6.cpp:19:28) 'void () noexcept' inline default trivial
```

When parsing the operator, the AST declares the original variable (this->data), however we should use a member variable (__this) from the lambda class, so we try to replace the operator expression in this PR.

Before:
```
operator() (c:@S@classB@F@foo#@Sa@F@operator()#1):
        // 30 file main6.cpp line 21 column 13 function operator()
        ASSIGN this->data=this->data + 1;                      
        // 31 file main6.cpp line 22 column 9 function operator()
        END_FUNCTION // operator()
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
failed: "data" is not a member of lambda
```

this PR:
```
operator() (c:@S@classB@F@foo#@Sa@F@operator()#1):
        // 30 file main6.cpp line 21 column 13 function operator()
        ASSIGN this->__this->data=this->__this->data + 1;
        // 31 file main6.cpp line 22 column 9 function operator()
        END_FUNCTION // operator()
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

I tried to be as robust as possible, but the AST provides very limited information, and I expected better methods.

### CODE REVIEW

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 1]
```diff
@@ -413,7 +414,27 @@ bool clang_cpp_convertert::get_struct_union_class_fields(
     if (annotate_class_field(*field, type, comp))
       return true;
 
+    if (auto cxxrd = llvm::dyn_cast<clang::CXXRecordDecl>(&rd))
```
Comment (lucasccordeiro) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp:419]: Can you add some documentation here?

File: src/clang-cpp-frontend/clang_cpp_adjust_expr.cpp [hunk 2]
```diff
@@ -59,6 +59,52 @@ void clang_cpp_adjust::gen_implicit_union_copy_move_constructor(symbolt &symbol)
...
+  const typet &class_symb = ns.follow(type.arguments()[0].type().subtype());
+
+  const struct_typet::componentst &components =
+    to_struct_type(class_symb).components();
+
+  for (const auto &c : components)
+  {
+    if (c.get_bool("#capture_this"))
+    {
+      //releace 'this->data' with 'this->__this->data'
```
Comment (lucasccordeiro) [inline src/clang-cpp-frontend/clang_cpp_adjust_expr.cpp]: Can you improve this description?

**Review by lucasccordeiro — APPROVED** [review]

**Comment by rafaelsamenezes** [comment]
Could you also add other tests cases? Such as: empty capture, without classes, with capture beyond of local and global vars, and with function parameters.

**Comment by XLiZHI** [comment]
Hi @rafaelsamenezes, we already have several such TCs:
https://github.com/XLiZHI/esbmc/tree/master/regression/esbmc-cpp11/lambda

**Review by rafaelsamenezes — APPROVED** [review]

**Comment by lucasccordeiro** [comment]
Thanks for submitting this PR, @XLiZHI.
